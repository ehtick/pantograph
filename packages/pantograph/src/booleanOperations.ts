import {
  cutFiguresLists,
  fuseFiguresLists,
  intersectFiguresLists,
} from "./algorithms/boolean/figureBooleans";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { listOfFigures } from "./utils/listOfFigures";

export function fuse(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    fuseFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function fuseAll(shapes: (Diagram | Figure | Loop)[]): Diagram {
  return shapes.reduce(
    (acc: Diagram, shape: Diagram | Figure | Loop) => fuse(acc, shape),
    new Diagram()
  );
}

export function cut(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    cutFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function intersect(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    intersectFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}
