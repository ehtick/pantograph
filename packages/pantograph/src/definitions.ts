export type Vector = [number, number];

export function isVector(vector: unknown): vector is Vector {
  return Array.isArray(vector) && vector.length === 2;
}
