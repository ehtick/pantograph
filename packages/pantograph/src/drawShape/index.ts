export { drawCircle } from "./drawCircle.js";
export { drawRect } from "./drawRect.js";
export { drawEllipse } from "./drawEllipse.js";
export { drawSVGPath } from "./drawSVGPath.js";
