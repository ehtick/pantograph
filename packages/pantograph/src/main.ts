import type { Vector } from "./definitions";
export { Vector };

export const DEG2RAD = Math.PI / 180;
export const RAD2DEG = 180 / Math.PI;

export function polarToCartesian(r: number, theta: number): Vector {
  const x = Math.cos(theta * DEG2RAD) * r;
  const y = Math.sin(theta * DEG2RAD) * r;
  return [x, y];
}

export function cartesianToPolar([x, y]: Vector): [number, number] {
  const r = Math.sqrt(x * x + y * y);
  const theta = Math.atan2(y, x) * RAD2DEG;
  return [r, theta];
}

export type {
  Diagram,
  Figure,
  Loop,
  Strand,
  Stroke,
  TransformationMatrix,
  BoundingBox,
  Segment,
  Line,
  Arc,
} from "./models/exports";

export { draw } from "./draw";
export {
  // Surface booleans
  fuseAll,
  fuse,
  cut,
  intersect,
  // Strand booleans
  eraseStrand,
  confineStrand,
  // Offset
  offset,
} from "./operations";

export { exportSVG, svgBody } from "./export/svg/exportSVG";

export { exportJSON } from "./export/json/exportJSON";
export { importJSON } from "./import/json/importJSON";
