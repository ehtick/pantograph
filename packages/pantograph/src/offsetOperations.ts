import {
  offsetFigures,
  outlineStrokeFigures,
} from "./algorithms/offsets/offsetFigure";
import { outlineStrand } from "./algorithms/offsets/offsetStroke.js";
import { Strand } from "./models/Strand.js";
import { Diagram } from "./models/Diagram.js";
import { Figure } from "./models/Figure.js";
import { Loop } from "./models/Loop.js";
import { listOfFigures } from "./utils/listOfFigures.js";

export function offset(
  shape: Diagram | Figure | Loop,
  offsetDistance: number
): Diagram {
  return offsetFigures(listOfFigures(shape), offsetDistance);
}

export function outlineStroke(
  shape: Diagram | Figure | Loop | Strand,
  outlineDistance: number,
  { endCap = "round" }: { endCap?: "butt" | "round" } = {}
): Diagram {
  if (shape instanceof Strand) {
    return outlineStrand(shape, outlineDistance, endCap);
  }
  return outlineStrokeFigures(listOfFigures(shape), outlineDistance);
}
