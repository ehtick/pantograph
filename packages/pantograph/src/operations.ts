import {
  cutFiguresLists,
  fuseFiguresLists,
  intersectFiguresLists,
} from "./algorithms/boolean/figureBooleans";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";

function listOfFigures(shape: Diagram | Figure | Loop): Figure[] {
  if (shape instanceof Figure) {
    return [shape];
  } else if (shape instanceof Loop) {
    return [new Figure(shape)];
  } else if (shape instanceof Diagram) {
    return shape.figures;
  }
  throw new Error("Unknown shape");
}

export function fuse(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    fuseFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function fuseAll(shapes: (Diagram | Figure | Loop)[]): Diagram {
  return shapes.reduce(
    (acc: Diagram, shape: Diagram | Figure | Loop) => fuse(acc, shape),
    new Diagram()
  );
}

export function cut(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    cutFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function intersect(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    intersectFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}
