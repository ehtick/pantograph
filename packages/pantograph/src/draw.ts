import { Vector } from "./definitions";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { Line } from "./models/segments/Line";
import { Segment } from "./models/segments/Segment";
import { TransformationMatrix } from "./models/TransformationMatrix";
import {
  polarToCartesian,
  DEG2RAD,
  subtract,
  sameVector,
} from "./vectorOperations";

function closeSegments(segments: Segment[]) {
  if (!segments.length) throw new Error("No segments to close");
  const firstSegment = segments[0];
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const lastSegment = segments.at(-1)!;

  if (sameVector(firstSegment.firstPoint, lastSegment.lastPoint))
    return segments;
  return [
    ...segments,
    new Line(lastSegment.lastPoint, firstSegment.firstPoint),
  ];
}

function loopySegmentsToDiagram(segments: Segment[]) {
  // Here we will need to do our best to fix cases where the drawing is
  // broken in some way (i.e. self-intersecting loops)

  return new Diagram([new Figure(new Loop(segments))]);
}

export class DrawingPen {
  protected pointer: Vector;
  protected firstPoint: Vector;
  protected pendingSegments: Segment[];

  constructor(origin: Vector = [0, 0]) {
    this.pointer = origin;
    this.firstPoint = origin;

    this.pendingSegments = [];
  }

  movePointerTo(point: Vector): this {
    if (this.pendingSegments.length)
      throw new Error(
        "You can only move the pointer if there is no segment defined"
      );

    this.pointer = point;
    this.firstPoint = point;
    return this;
  }

  protected saveSegment(segment: Segment) {
    this.pendingSegments.push(segment);
    return this;
  }

  lineTo(point: Vector): this {
    const segment = new Line(this.pointer, point);
    this.pointer = point;
    return this.saveSegment(segment);
  }

  line(xDist: number, yDist: number): this {
    return this.lineTo([this.pointer[0] + xDist, this.pointer[1] + yDist]);
  }

  vLine(distance: number): this {
    return this.line(0, distance);
  }

  hLine(distance: number): this {
    return this.line(distance, 0);
  }

  vLineTo(yPos: number): this {
    return this.lineTo([this.pointer[0], yPos]);
  }

  hLineTo(xPos: number): this {
    return this.lineTo([xPos, this.pointer[1]]);
  }

  polarLineTo([r, theta]: Vector): this {
    const angleInRads = theta * DEG2RAD;
    const point = polarToCartesian(r, angleInRads);
    return this.lineTo(point);
  }

  polarLine(distance: number, angle: number): this {
    const angleInRads = angle * DEG2RAD;
    const [x, y] = polarToCartesian(distance, angleInRads);
    return this.line(x, y);
  }

  tangentLine(distance: number): this {
    const previousCurve = this.pendingSegments.at(-1);

    if (!previousCurve)
      throw new Error("You need a previous segment to sketch a tangent line");

    const [xDir, yDir] = previousCurve.tangentAtLastPoint;
    return this.line(xDir * distance, yDir * distance);
  }

  close(): Diagram {
    const segments = closeSegments(this.pendingSegments);
    return loopySegmentsToDiagram(segments);
  }

  closeWithMirror(): Diagram {
    if (!this.pendingSegments.length) throw new Error("No segments to close");

    const firstSegment = this.pendingSegments[0];
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastSegment = this.pendingSegments.at(-1)!;

    const mirrorVector = subtract(
      lastSegment.lastPoint,
      firstSegment.firstPoint
    );
    const mirrorTranform = new TransformationMatrix().mirrorLine(
      mirrorVector,
      firstSegment.firstPoint
    );

    const mirroredSegments = this.pendingSegments.map((segment) =>
      segment.transform(mirrorTranform).reverse()
    );
    mirroredSegments.reverse();

    return loopySegmentsToDiagram([
      ...this.pendingSegments,
      ...mirroredSegments,
    ]);
  }
}

export function draw(origin: Vector = [0, 0]): DrawingPen {
  return new DrawingPen(origin);
}
