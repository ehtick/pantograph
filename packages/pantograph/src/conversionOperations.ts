import { Segment } from "./models/segments/Segment";
import { Arc } from "./models/segments/Arc";
import { Line } from "./models/segments/Line";
import { QuadraticBezier } from "./models/segments/QuadraticBezier";
import { CubicBezier } from "./models/segments/CubicBezier";
import { EllipseArc } from "./models/segments/EllipseArc";
import { approximateBezierCurveAsArcs } from "./algorithms/conversions/bezierToArcs";
import { approximateEllipticalArcAsCubicBeziers } from "./algorithms/conversions/ellipseToBezier";
import { genericConversion } from "./algorithms/conversions/helpers";
import { Diagram, Figure, Loop, Strand } from "./main";

function mapAsArcsAndLines(s: Segment, tolerance = 1e-3): Segment[] {
  if (s instanceof Arc || s instanceof Line) {
    return [s];
  }
  if (s instanceof QuadraticBezier || s instanceof CubicBezier) {
    return approximateBezierCurveAsArcs(s, tolerance);
  }
  if (s instanceof EllipseArc) {
    return approximateEllipticalArcAsCubicBeziers(s).flatMap((s) =>
      approximateBezierCurveAsArcs(s, tolerance),
    );
  }

  throw new Error(`Unknown segment type: ${s}`);
}

function convertToArcsAndLines(shape: Diagram): Diagram;
function convertToArcsAndLines(shape: Figure): Figure;
function convertToArcsAndLines(shape: Loop): Loop;
function convertToArcsAndLines(shape: Strand): Strand;
function convertToArcsAndLines(shape: Segment): Segment[];
function convertToArcsAndLines(
  shape: Diagram | Figure | Loop | Strand | Segment,
  tolerance = 1e-3,
) {
  const mapper = (segment: Segment) => mapAsArcsAndLines(segment, tolerance);
  return genericConversion(shape, mapper);
}

export { convertToArcsAndLines };
