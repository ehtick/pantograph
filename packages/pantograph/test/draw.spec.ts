import { describe, it, expect } from "vitest";
import { Vector } from "../src/definitions";
import { draw } from "../src/draw";
//import { debugImg } from "./debug";

import { rect, polygon } from "./quickShapes";

const POLYGON: Vector[] = [
  [-2, 3],
  [-1, -3],
  [2, -2],
  [3, 2],
];

describe("draw", () => {
  it("should work with absolute coordinates", () => {
    const fig = draw(POLYGON[0])
      .lineTo(POLYGON[1])
      .lineTo(POLYGON[2])
      .lineTo(POLYGON[3])
      .lineTo(POLYGON[0])
      .close();

    expect(fig.figures[0].contour).toBeLoop(polygon(POLYGON));
  });

  it("should work horizonal and vertical relative coordinates", () => {
    const fig = draw([-2, -2]).hLine(4).vLine(4).hLine(-4).vLine(-4).close();
    expect(fig.figures[0].contour).toBeLoop(rect(4, 4));
  });

  it("should close when missing a last link", () => {
    const fig = draw(POLYGON[0])
      .lineTo(POLYGON[1])
      .lineTo(POLYGON[2])
      .lineTo(POLYGON[3])
      .close();

    expect(fig.figures[0].contour).toBeLoop(polygon(POLYGON));
  });

  it("should close with a mirror", () => {
    const fig = draw(POLYGON[0])
      .lineTo(POLYGON[1])
      .lineTo(POLYGON[2])
      .closeWithMirror();

    expect(fig.figures[0].contour).toBeLoop(
      polygon([
        [-2, 3],
        [-1, -3],
        [2, -2],
        [3.6341463414634148, 0.7073170731707324],
      ]),
    );
  });

  it("should draw half ellipses as expected", () => {
    const fig = draw([0, 0]).halfEllipseTo([2, 2], 0.5).vLine(-2).close();
    //debugImg(fig);
    expect(fig).toMatchSnapshot();

    const fig2 = draw([0, 0]).halfEllipseTo([2, 2], 5).vLine(-2).close();
    //debugImg(fig2);
    expect(fig2).toMatchSnapshot();
  });

  it("should close with a half circle without issue ", () => {
    const fig = draw([0, 0]).bulgeArcTo([2, 2], 1).close();
    //debugImg(fig);
    expect(fig).toMatchSnapshot();
  });
});
