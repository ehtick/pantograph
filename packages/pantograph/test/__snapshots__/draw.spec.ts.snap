// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`draw > should close with a half circle without issue  1`] = `
Diagram {
  "_boundingBox": null,
  "figures": [
    Figure {
      "contour": Loop {
        "_boundingBox": null,
        "_clockwise": null,
        "segments": [
          Arc {
            "_angularLength": 3.141592653589793,
            "_boundingBox": BoundingBox {
              "xMax": 2.414213563373095,
              "xMin": -1e-9,
              "yMax": 2.000000001,
              "yMin": -0.41421356337309523,
            },
            "_coefficients": null,
            "_firstAngle": 3.9269908169872414,
            "_lastAngle": 0.7853981633974483,
            "_radius": 1.4142135623730951,
            "center": [
              1,
              1,
            ],
            "clockwise": false,
            "firstPoint": [
              0,
              0,
            ],
            "lastPoint": [
              2,
              2,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Line {
            "_V": [
              -2,
              -2,
            ],
            "_boundingBox": BoundingBox {
              "xMax": 2.000000001,
              "xMin": -1e-9,
              "yMax": 2.000000001,
              "yMin": -1e-9,
            },
            "_slope": null,
            "_yIntercept": null,
            "firstPoint": [
              2,
              2,
            ],
            "lastPoint": [
              0,
              0,
            ],
            "precision": 1e-9,
            "segmentType": "LINE",
          },
        ],
        "strokeType": "LOOP",
      },
      "holes": [],
    },
  ],
}
`;

exports[`draw > should draw half ellipses as expected 1`] = `
Diagram {
  "_boundingBox": null,
  "figures": [
    Figure {
      "contour": Loop {
        "_boundingBox": null,
        "_clockwise": null,
        "segments": [
          EllipseArc {
            "_boundingBox": BoundingBox {
              "xMax": 2,
              "xMin": -0.06066017923040201,
              "yMax": 2.0606601792304016,
              "yMin": 0,
            },
            "_coefficients": null,
            "_deltaAngle": 3.1415926957366413,
            "_ellipseReferenceFrameTransform": TransformationMatrix {
              "_matrix": [
                0.7071067811865476,
                0.7071067811865475,
                -1.414213562373095,
                -0.7071067811865475,
                0.7071067811865476,
                -1.0536712280462268e-8,
                0,
                0,
                1,
              ],
            },
            "_firstAngle": 3.1415926746632175,
            "_focals": [
              [
                1.9354143392429048,
                1.9354143541440658,
              ],
              [
                0.06458564585593396,
                0.06458566075709526,
              ],
            ],
            "_lastAngle": 6.283185286106162,
            "_linearExentricity": 1.3228756555322954,
            "_reverseEllipseReferenceFrameTransform": TransformationMatrix {
              "_matrix": [
                0.7071067811865476,
                -0.7071067811865475,
                0.9999999925494194,
                0.7071067811865475,
                0.7071067811865476,
                1.0000000074505806,
                0,
                0,
                1,
              ],
            },
            "center": [
              0.9999999925494194,
              1.0000000074505806,
            ],
            "clockwise": true,
            "firstPoint": [
              0,
              0,
            ],
            "lastPoint": [
              2,
              2,
            ],
            "majorRadius": 1.4142135623730951,
            "minorRadius": 0.5,
            "precision": 0.000001,
            "segmentType": "ELLIPSE_ARC",
            "tiltAngle": 0.7853981633974483,
          },
          Line {
            "_V": [
              0,
              -2,
            ],
            "_boundingBox": BoundingBox {
              "xMax": 2.000000001,
              "xMin": 1.999999999,
              "yMax": 2.000000001,
              "yMin": -1e-9,
            },
            "_slope": null,
            "_yIntercept": null,
            "firstPoint": [
              2,
              2,
            ],
            "lastPoint": [
              2,
              0,
            ],
            "precision": 1e-9,
            "segmentType": "LINE",
          },
          Line {
            "_V": [
              -2,
              0,
            ],
            "_boundingBox": BoundingBox {
              "xMax": 2.000000001,
              "xMin": -1e-9,
              "yMax": 1e-9,
              "yMin": -1e-9,
            },
            "_slope": null,
            "_yIntercept": null,
            "firstPoint": [
              2,
              0,
            ],
            "lastPoint": [
              0,
              0,
            ],
            "precision": 1e-9,
            "segmentType": "LINE",
          },
        ],
        "strokeType": "LOOP",
      },
      "holes": [],
    },
  ],
}
`;

exports[`draw > should draw half ellipses as expected 2`] = `
Diagram {
  "_boundingBox": null,
  "figures": [
    Figure {
      "contour": Loop {
        "_boundingBox": null,
        "_clockwise": null,
        "segments": [
          EllipseArc {
            "_boundingBox": BoundingBox {
              "xMax": 2,
              "xMin": -2.674234673779411,
              "yMax": 4.674234673779411,
              "yMin": 0,
            },
            "_coefficients": null,
            "_deltaAngle": 3.1415926873072717,
            "_ellipseReferenceFrameTransform": TransformationMatrix {
              "_matrix": [
                -0.7071067811865475,
                0.7071067811865476,
                -8.429369702245282e-8,
                -0.7071067811865476,
                -0.7071067811865475,
                1.414213562373095,
                0,
                0,
                1,
              ],
            },
            "_firstAngle": 1.570796343653636,
            "_focals": [
              [
                -2.3911650511672784,
                4.391165051167279,
              ],
              [
                4.391164931957989,
                -2.3911649319579893,
              ],
            ],
            "_lastAngle": 4.71238896352595,
            "_linearExentricity": 4.795831523312719,
            "_reverseEllipseReferenceFrameTransform": TransformationMatrix {
              "_matrix": [
                -0.7071067811865475,
                -0.7071067811865476,
                0.9999999403953552,
                0.7071067811865476,
                -0.7071067811865475,
                1.0000000596046448,
                0,
                0,
                1,
              ],
            },
            "center": [
              0.9999999403953552,
              1.0000000596046448,
            ],
            "clockwise": true,
            "firstPoint": [
              0,
              0,
            ],
            "lastPoint": [
              2,
              2,
            ],
            "majorRadius": 5,
            "minorRadius": 1.4142135623730951,
            "precision": 0.000001,
            "segmentType": "ELLIPSE_ARC",
            "tiltAngle": 2.356194490192345,
          },
          Line {
            "_V": [
              0,
              -2,
            ],
            "_boundingBox": BoundingBox {
              "xMax": 2.000000001,
              "xMin": 1.999999999,
              "yMax": 2.000000001,
              "yMin": -1e-9,
            },
            "_slope": null,
            "_yIntercept": null,
            "firstPoint": [
              2,
              2,
            ],
            "lastPoint": [
              2,
              0,
            ],
            "precision": 1e-9,
            "segmentType": "LINE",
          },
          Line {
            "_V": [
              -2,
              0,
            ],
            "_boundingBox": BoundingBox {
              "xMax": 2.000000001,
              "xMin": -1e-9,
              "yMax": 1e-9,
              "yMin": -1e-9,
            },
            "_slope": null,
            "_yIntercept": null,
            "firstPoint": [
              2,
              0,
            ],
            "lastPoint": [
              0,
              0,
            ],
            "precision": 1e-9,
            "segmentType": "LINE",
          },
        ],
        "strokeType": "LOOP",
      },
      "holes": [],
    },
  ],
}
`;
